### PR TITLE
[BE20-246] Sendgrid API async client

### PR DESCRIPTION
**Pull request by k-vinogradov**

### CODE REVIEW

File: sendgrid/__init__.py
Comment (antonmolodykh) [inline sendgrid/__init__.py]: Add end line please

File: sendgrid/async_sendgrid.py
Comment (antonmolodykh) [inline sendgrid/async_sendgrid.py]: One of them is redundant

File: sendgrid/async_sendgrid.py
Comment (antonmolodykh) [inline sendgrid/async_sendgrid.py]: Do not use personal mail here

File: sendgrid/sendgrid.py
Comment (antonmolodykh) [inline sendgrid/sendgrid.py]: Extra spaces

**Review by antonmolodykh — CHANGES_REQUESTED** [review]

File: sendgrid/sendgrid.py
Comment (k-vinogradov) [reply sendgrid/sendgrid.py]: It's a foreign code. We avoid any unnecessary changes in a foreign code (including code style fixing)

Comment (k-vinogradov) [reply sendgrid/sendgrid.py]: Actually this line is not. Fixed.

**Review by antonmolodykh — APPROVED** [review]

File: test/requirements.txt
Comment (DrLoboto) [inline test/requirements.txt]: Don't see `pytest` usage. `asynctest` is async tests dependency only.

File: tox.ini
Comment (DrLoboto) [inline tox.ini]: Why `--ignore-errors`?

File: tox.ini
Comment (DrLoboto) [inline tox.ini]: Better be covered by `extras` setting in async envs.

File: tox.ini
Comment (DrLoboto) [inline tox.ini:18]: Looks like original decision was "provide key settings explicitly even if they are inherited".

File: tox.ini
Comment (DrLoboto) [inline tox.ini]: Coverage is gathered but not used. To get full coverage easy way there should be single run over all needed files. Manual report run of last env pass shown 82% on `sendgrid.async_sendgrid`.

File: tox.ini
Comment (DrLoboto) [inline tox.ini]: Why `py34` is excluded?

**Review by DrLoboto — CHANGES_REQUESTED** [review]
Didn't check tests code, only run them.

File: tox.ini
Comment (k-vinogradov) [reply tox.ini]: Sendgrid doesn't support Python 3.4. Please see the `setup.py`:

```python
python_requires='>=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*, !=3.4.*',
```

But it's a good notice because I missed Python 3.8 too

File: test/requirements.txt
Comment (k-vinogradov) [reply test/requirements.txt]: The file isn't used now. Removed both of lines.

File: tox.ini
Comment (k-vinogradov) [reply tox.ini]: Without `--ignore-errors` `coverage` for Python 2.7 fails because of `invalid syntax`. Extracted to the Python 2.7 env.

File: tox.ini
Comment (k-vinogradov) [reply tox.ini]: `extras` requires adding the test dependencies to the `setup.py`. It doesn't match the maintainer's approach. Actually the entire `tox.ini` can be replaced with the following lines to avoid lots of env sections and default params redefining.
```
[tox]
envlist = py27, py35, py36, py37, py38, {py35, py36, py37, py38}-async

[testenv]
commands = coverage erase
           !async: coverage run -m unittest discover -p "test_*.py" []
           async:  coverage run -m unittest discover -p "*test_*.py" []
           !py27:  coverage report
           py27:   coverage report --ignore-errors

deps = -rrequirements.txt
       coverage
       async: asynctest
       async: python-http-client[async]==3.2.6dev5
       py27: mock
```

File: tox.ini
Comment (k-vinogradov) [reply tox.ini:18]: Removed redundant sections and default redefining from the config

File: tox.ini
Comment (k-vinogradov) [reply tox.ini]: Joined to the single run for the async envs.

**Review by DrLoboto — COMMENTED** [review]
Please compact tox configs as you proposed, but instead of conditional lines envs sections usage is clearer. Remove coverage errors ignore in Python 2.7 env and exclude async file in non-async envs with `coverage report --omit="sendgrid/async_sendgrid.py"`

File: tox.ini
Comment (DrLoboto) [inline tox.ini]: It must be omit in non-async tests only. The fix will make `commands` section in `py27` env obsolete.

File: tox.ini
Comment (DrLoboto) [inline tox.ini]: If `commands` go out this may also be set as conditional dependency in main section.

**Review by DrLoboto — CHANGES_REQUESTED** [review]

**Review by DrLoboto — COMMENTED** [review]
All Python3 tests run finish with:
```
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name='/dev/null' mode='w' encoding='UTF-8'>
```
Can it be fixed? It is pure tests flaw or main code one?

**Comment by k-vinogradov** [comment]
> Can it be fixed? It is pure tests flaw or main code one?

It's an original test code. They [open](https://github.com/savonixinc/sendgrid-python/blob/d02d15f20565054d8be600976b514b2b95b753dc/test/test_sendgrid.py#L20) `/dev/null` descriptor. I have no idea why. It doesn't seem they use it.

I didn't remove it just in case. Just added [tear down](https://github.com/savonixinc/sendgrid-python/blob/d02d15f20565054d8be600976b514b2b95b753dc/test/test_sendgrid.py#L24).

**Review by DrLoboto — APPROVED** [review]
PR is ready to go. Do not merge yet - will experiment on Jenkins jobs targeted to PRs, not individual pushes.

**Pull request merged**
